### ISSUE TITLE
Crash: Cannot destructure property 'message' of 'testStepResult' as it is undefined

### ISSUE THREAD

**Issue by gpepignosis** [badeball/cypress-cucumber-preprocessor#1099]
### Current behavior

This bug does NOT happen every time.
This bug has not yet been replicated locally and only has appeared in a NodeJS environment inside a Github action, but more likely that Github action does not play a role here. Rather more likely that this happens when executing the full project, all the feature files.

### Desired behavior

Not to yield this crash

### Test code to reproduce

Sorry as explained above is difficult to reproduce and the scenario that it happened has pretty straightforward cypress code.
The only code that is let's say out of the ordinary is using this cypress command:
```
Cypress.Commands.add('downloadByClick',
  {prevSubject: false},
  (subject, wait_seconds, download_name) => {
    cy.window().document().then(function (doc) {
        doc.addEventListener('click', () => {
          setTimeout(function () { doc.location.reload() }, wait_seconds * 1000)
        });
        cy.get(subject).should('be.visible').click();
    });
    const downloadsFolder = Cypress.config("downloadsFolder");
    return cy.readFile(path.join(downloadsFolder, download_name));
  }
)
```

### Versions

* **Cypress version**: 13.1.0
* **Preprocessor version**: 18.0.5
* **Node version**: 18.6.0

### Checklist

- [X] I've read the [FAQ](https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/docs/faq.md).
- [X] I've read [instructions for logging issues](https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I'm not using `cypress-cucumber-preprocessor@4.3.1` (package name has changed and it is no longer the most recent version, see [#689](https://github.com/badeball/cypress-cucumber-preprocessor/issues/689)).
No we instead are using: `"@badeball/cypress-cucumber-preprocessor": "^18.0.5",`

### FULL REPORT OF THE ERROR BELOW

```
    (Attempt 1 of 3) Downloading a sample Excel file
    (Attempt 1 of 3) Downloading a sample Excel file
    (Attempt 2 of 3) Downloading a sample Excel file
    (Attempt 2 of 3) Downloading a sample Excel file
    1) Downloading a sample Excel file
    2) "after each" hook for "Downloading a sample Excel file"


  14 passing (3m)
  2 failing

  1) Import
       Downloading a sample Excel file:
     CypressError: `cy.task('cypress-cucumber-preprocessor:test-case-started')` failed with the following error:

> Unexpected state in testCaseStartedHandler: step-finished (this might be a bug, please report at https://github.com/badeball/cypress-cucumber-preprocessor)

https://on.cypress.io/api/task
      at <unknown> (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:135258:70)
      at tryCatcher (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1807:23)
      at Promise._settlePromiseFromHandler (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1519:31)
      at Promise._settlePromise (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1576:18)
      at Promise._settlePromise0 (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1621:10)
      at Promise._settlePromises (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1697:18)
      at _drainQueueStep (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2407:12)
      at _drainQueue (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2400:9)
      at Async._drainQueues (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2416:5)
      at Async.drainQueues (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2286:14)
  From Your Spec Code:
      at taskTestCaseStarted (https://qa1basic.bleeding8.talentlms.com/__cypress/tests?p=cypress/e2e/talentlms/import-export/import.feature:14044:14)
      at Context.eval (https://qa1basic.bleeding8.talentlms.com/__cypress/tests?p=cypress/e2e/talentlms/import-export/import.feature:14203:11)
  
  From Node.js Internals:
    Error: Unexpected state in testCaseStartedHandler: step-finished (this might be a bug, please report at https://github.com/badeball/cypress-cucumber-preprocessor)
        at createError (/home/runner/work/tlms-plus-ui-tests/tlms-plus-ui-tests/node_modules/@badeball/cypress-cucumber-preprocessor/dist/helpers/error.js:9:12)
        at testCaseStartedHandler (/home/runner/work/tlms-plus-ui-tests/tlms-plus-ui-tests/node_modules/@badeball/cypress-cucumber-preprocessor/dist/plugin-event-handlers.js:325:43)
        at invoke (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:246:16)
        at <unknown> (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:59:14)
        at tryCatcher (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/bluebird/js/release/util.js:16:23)
        at Promise.attempt.Promise.try (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/bluebird/js/release/method.js:39:29)
        at Object.wrapChildPromise (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:58:23)
        at RunPlugins.taskExecute (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:252:10)
        at RunPlugins.execute (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:166:21)
        at EventEmitter.<anonymous> (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:56:12)
        at EventEmitter.emit (node:events:514:28)
        at process.<anonymous> (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:33:22)
        at process.emit (node:events:514:28)
        at emit (node:internal/child_process:937:14)
        at process.processTicksAndRejections (node:internal/process/task_queues:83:21)

  2) Import
       "after each" hook for "Downloading a sample Excel file":
     CypressError: `cy.task('cypress-cucumber-preprocessor:test-step-finished')` failed with the following error:

> Unexpected state in testStepFinishedHandler: step-finished (this might be a bug, please report at https://github.com/badeball/cypress-cucumber-preprocessor)

https://on.cypress.io/api/task

Because this error occurred during a `after each` hook we are skipping the remaining tests in the current suite: `Import`
      at <unknown> (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:135258:70)
      at tryCatcher (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1807:23)
      at Promise._settlePromiseFromHandler (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1519:31)
      at Promise._settlePromise (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1576:18)
      at Promise._settlePromise0 (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1621:10)
      at Promise._settlePromises (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:1697:18)
      at _drainQueueStep (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2407:12)
      at _drainQueue (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2400:9)
      at Async._drainQueues (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2416:5)
      at Async.drainQueues (https://qa1basic.bleeding8.talentlms.com/__cypress/runner/cypress_runner.js:2286:14)
  From Your Spec Code:
      at taskTestStepFinished (https://qa1basic.bleeding8.talentlms.com/__cypress/tests?p=cypress/e2e/talentlms/import-export/import.feature:14065:14)
      at Context.afterEachHandler (https://qa1basic.bleeding8.talentlms.com/__cypress/tests?p=cypress/e2e/talentlms/import-export/import.feature:14460:13)
      at Context.eval (https://qa1basic.bleeding8.talentlms.com/__cypress/tests?p=cypress/e2e/talentlms/import-export/import.feature:14104:30)
  
  From Node.js Internals:
    Error: Unexpected state in testStepFinishedHandler: step-finished (this might be a bug, please report at https://github.com/badeball/cypress-cucumber-preprocessor)
        at createError (/home/runner/work/tlms-plus-ui-tests/tlms-plus-ui-tests/node_modules/@badeball/cypress-cucumber-preprocessor/dist/helpers/error.js:9:12)
        at testStepFinishedHandler (/home/runner/work/tlms-plus-ui-tests/tlms-plus-ui-tests/node_modules/@badeball/cypress-cucumber-preprocessor/dist/plugin-event-handlers.js:380:43)
        at invoke (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:246:16)
        at <unknown> (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:59:14)
        at tryCatcher (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/bluebird/js/release/util.js:16:23)
        at Promise.attempt.Promise.try (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/bluebird/js/release/method.js:39:29)
        at Object.wrapChildPromise (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:58:23)
        at RunPlugins.taskExecute (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:252:10)
        at RunPlugins.execute (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:166:21)
        at EventEmitter.<anonymous> (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:56:12)
        at EventEmitter.emit (node:events:514:28)
        at process.<anonymous> (/home/runner/.cache/Cypress/13.1.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:33:22)
        at process.emit (node:events:514:28)
        at emit (node:internal/child_process:937:14)
        at process.processTicksAndRejections (node:internal/process/task_queues:83:21)


```

![image](https://github.com/badeball/cypress-cucumber-preprocessor/assets/127388253/67cdff8f-c9e0-4ded-9b74-51bf7c300dd0)

**Comment by badeball** [comment]
Can you provide a runable example that illustrates how `downloadByClick` is used?

**Comment by gpepignosis** [comment]
Hi @badeball we were hoping all the rest of the logs would provide useful insight for guessing what might have gone wrong.

If this will be of any help the download by click can be used in any cypress scenario like so:

`cy.downloadByClick("some css selector to button", 5, "hello.bin").should('not.be.empty');`

**Comment by badeball** [comment]
This doesn't help me, no. Why are you reloading the page upon click? Does the intermittent failures happen without the reload?

**Comment by gpepignosis** [comment]
Hi @badeball you can read this stackoverflow to understand the issue of Cypress and how the code above workarounds it https://stackoverflow.com/questions/72782444/wait-for-page-to-load-after-click-on-download-button
Hope this gives you some extra context. Looking forward to your fix

And @badeball another guess here is that we are using Cypress retries as described here: https://docs.cypress.io/guides/guides/test-retries
This might play a role on this issue. Hope this also helps.

**Comment by badeball** [comment]
I can't reproduce your issue with this information. You need to provide a reproducible example, at least something which you've seen fail in CI.

**Comment by gpepignosis** [comment]
@badeball The only thing you might be able to do on your side is to simply make the code more robust/simplified on your side, if applicable, and this might resolve itself without getting to the bottom of it.

On our side we will try to replicate it consistently. Give us some time

**Comment by gpepignosis** [comment]
@badeball We are afraid this is difficult to replicate outside of the github action, and happens inconsistency in some nightly runs, when the entire suite is executed.
One could try a range of scenarios with the @wip tag which are a bit flaky, so you could have at least a retry, similar to this:
`npx cypress-repeat run -n 10 --browser chrome --config baseUrl="https://xxx.talentlms.com",retries=2 -e TAGS="@wip" --spec="" --headless`

What is your best guess(es) of what might have caused this crash on the plugin side ?
It might help us to reproduce it more consistently

**Comment by badeball** [comment]
> One could try a range of scenarios with the https://github.com/wip tag which are a bit flaky, so you could have at least a retry, similar to this:
> `npx cypress-repeat run -n 10 --browser chrome --config baseUrl="https://xxx.talentlms.com",retries=2 -e TAGS="@wip" --spec="" --headless`

What is the `cypress-repeat` executable you're referring to here?

> What is your best guess(es) of what might have caused this crash on the plugin side ?

Based on your posted output:

```
    (Attempt 1 of 3) Downloading a sample Excel file
    (Attempt 1 of 3) Downloading a sample Excel file
    (Attempt 2 of 3) Downloading a sample Excel file
    (Attempt 2 of 3) Downloading a sample Excel file
    1) Downloading a sample Excel file
    2) "after each" hook for "Downloading a sample Excel file"
```

.. it almost seems as if the two tests are executed somewhat in parallel, as this is not the output I expect. Rather, I expect to see something like

```
    (Attempt 1 of 3) Downloading a sample Excel file
    (Attempt 2 of 3) Downloading a sample Excel file
    1) Downloading a sample Excel file
    (Attempt 1 of 3) Downloading a sample Excel file
    (Attempt 2 of 3) Downloading a sample Excel file
    2) "after each" hook for "Downloading a sample Excel file"
```

If two tests are infact interlaced, then something outside of this plugin is wrong.

**Comment by gpepignosis** [comment]
Hi @badeball 
cypress repeat is a very simple plugin which is found here: https://github.com/bahmutov/cypress-repeat 

Your observations make sense, and if this is not working as expected running tests in parallel then this is either a bug in Cypress or perhaps it might be a bug having Cypress with retries and Badeball plugin. I believe this you can replicate quite easy on your side to notice if this is happening

**Comment by badeball** [comment]
> I believe this you can replicate quite easy on your side to notice if this is happening

I don’t make it a habit to chase down problems in other tools. If you think this is easy to reproduce, then you need to provide me with a reproducible example.

**Comment by gpepignosis** [comment]
You don't have the curse of QA engineers!
Anyway we tried this: 
```
    if(Cypress.currentRetry == 0 || Cypress.currentRetry == 1){
        expect(true).to.be.equal(false);
    }
```
but could not reproduce it. Happens only in Github action

**Comment by Nandu890** [comment]
Hi @badeball, I also encountered the same issue while executing the spec in the terminal. I observed that it only occurs when we include additional options like ("html," "messages," "json") in "cypress-cucumber-preprocessor," except for "stepDefinitions" in package,json.
![Screenshot 2023-10-06 213834](https://github.com/badeball/cypress-cucumber-preprocessor/assets/72277827/8fbbf75d-b42a-4e72-916a-5b855cb3c6e3)

**Comment by badeball** [comment]
@Nandu890, you did not encounter the same issue. Furthermore, the answer to your issue is readily availabile in the documentation (read it). Lastly, if you pollute this issue page anymore, you'll be banned from the tracker altogheter.

**Comment by badeball** [comment]
Closing due to lack of a reproducible example. Feel free to open up another issue if anything changes in this regard.
